tools/clang/lib/SPIRV/GlPerVertex.h

```diff
@@ -0,0 +1,205 @@
+  /// we can caculate the ClipDistance/CullDistance array layout.
+  bool recordClipCullDistanceDecl(const DeclaratorDecl *decl, bool asInput);
+
+  /// Calculates the layout for ClipDistance/CullDistance arrays.
+  void calculateClipCullDistanceArraySize();
+
+  /// Emits SPIR-V code for the input and/or ouput gl_PerVertex builtin
+  /// interface blocks. If inputArrayLength is not zero, the input gl_PerVertex
+  /// will have an additional arrayness of the given size. Similarly for
+  /// outputArrayLength.
+  ///
+  /// Note that this method should be called after recordClipCullDistanceDecl()
+  /// and calculateClipCullDistanceArraySize().
+  void generateVars(uint32_t inputArrayLength, uint32_t outputArrayLength);
+
+  /// Returns the <result-id>s for stage input variables.
+  llvm::SmallVector<uint32_t, 4> getStageInVars() const;
+  /// Returns the <result-id>s for stage output variables.
+  llvm::SmallVector<uint32_t, 4> getStageOutVars() const;
+
+  /// Tries to access the builtin translated from the given HLSL semantic of
+  /// the given index. If sigPoint dicates this is input, builtins will be read
+  /// to compose a new temporary value of the correct type and writes to *value.
+  /// Otherwise, the *value will be decomposed and writes to the builtins.
+  /// Emits SPIR-V instructions and returns true if we are accessing builtins
+  /// belonging to gl_PerVertex. Does nothing and returns true if we are
+  /// accessing builtins not in gl_PerVertex. Returns false if errors occurs.
+  bool tryToAccess(hlsl::Semantic::Kind, uint32_t semanticIndex,
+                   llvm::Optional<uint32_t> invocationId, uint32_t *value,
+                   hlsl::SigPoint::Kind sigPoint);
+
+private:
+  template <unsigned N>
+  DiagnosticBuilder emitError(const char (&message)[N],
+                              SourceLocation loc = {}) {
+    const auto diagId = astContext.getDiagnostics().getCustomDiagID(
+        clang::DiagnosticsEngine::Error, message);
+    return astContext.getDiagnostics().Report(loc, diagId);
+  }
+
+  /// Creates a gl_PerVertex interface block variable. If arraySize is not zero,
+  /// The created variable will be an array of gl_PerVertex of the given size.
+  /// Otherwise, it will just be a plain struct.
+  uint32_t createBlockVar(bool asInput, uint32_t arraySize);
+  /// Creates a stand-alone Position builtin variable.
+  uint32_t createPositionVar(bool asInput);
+  /// Creates a stand-alone ClipDistance builtin variable.
+  uint32_t createClipDistanceVar(bool asInput, uint32_t arraySize);
+  /// Creates a stand-alone CullDistance builtin variable.
+  uint32_t createCullDistanceVar(bool asInput, uint32_t arraySize);
+
+  /// Emits SPIR-V instructions for reading the Position builtin.
+  uint32_t readPosition() const;
+  /// Emits SPIR-V instructions for reading the data starting from offset in
+  /// the ClipDistance/CullDistance builtin. For clipCullIndex, 2 means
```

Comment:
Since `clipCullIndex` is invalid for any value other than 2 and 3, I would change `uint32_t clipCullIndex` to `bool isClip`
Then inside the function:
```cpp
const bool clipCullIndex = isClip ? 2 : 3;
```

Reply:
similarly for `writeClipCullArrayFromType` below

Reply:
Done